[ wk2 Debug ] loader/stateobjects/pushstate-size-iframe.html is very slow, usually times out

cdumez opened the pull request:
#### 4e8b5c1a05ae111def3a5754f16af81ff19ae3a0
<pre>
[ wk2 Debug ] loader/stateobjects/pushstate-size-iframe.html is very slow, usually times out
<a href="https://bugs.webkit.org/show_bug.cgi?id=268836">https://bugs.webkit.org/show_bug.cgi?id=268836</a>
<a href="https://rdar.apple.com/122401579">rdar://122401579</a>

Reviewed by Ryosuke Niwa.

This test was slow, especially in debug because it had to reach the 64MB size limit by doing
repeated history.pushState() calls. This would obviously take a while. To address the issue,
I am introducing a new window.internals API to override this limit. In the test, we now
lower this limit to 4MB, which makes the test run a lot faster.

* LayoutTests/loader/stateobjects/pushstate-size-iframe-expected.txt:
* LayoutTests/loader/stateobjects/pushstate-size-iframe.html:
* LayoutTests/loader/stateobjects/resources/pushstate-iframe.html:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/page/History.cpp:
(WebCore::History::totalStateObjectPayloadLimit const):
(WebCore::History::stateObjectAdded):
* Source/WebCore/page/History.h:
* Source/WebCore/page/LocalDOMWindow.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::resetToConsistentState):
(WebCore::Internals::setHistoryTotalStateObjectPayloadLimitOverride):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: <a href="https://commits.webkit.org/275452@main">https://commits.webkit.org/275452@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/98870facc7db1afe96ae57fef9a5f56b8d0cef0b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41856 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20871 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44238 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44438 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37951 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/44163 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24026 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18201 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34595 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42430 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17800 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36046 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15266 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/15488 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/37069 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45835 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38044 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37391 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/41150 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16671 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13689 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/39598 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18290 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/36345 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9380 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18349 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17934 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/page/History.h:
```diff
@@ -51,6 +51,8 @@ class History final : public ScriptWrappable, public RefCounted<History>, public
     ExceptionOr<ScrollRestoration> scrollRestoration() const;
     ExceptionOr<void> setScrollRestoration(ScrollRestoration);
 
+    void setTotalStateObjectPayloadLimitOverride(std::optional<uint32_t> override) { m_totalStateObjectPayloadLimitOverride = override; }
```
I would have called this argument "limit" instead.

cdumez replied:
Will rename. Thanks.

cdumez committed:
[ wk2 Debug ] loader/stateobjects/pushstate-size-iframe.html is very slow, usually times out
https://bugs.webkit.org/show_bug.cgi?id=268836
rdar://122401579

Reviewed by Ryosuke Niwa.

This test was slow, especially in debug because it had to reach the 64MB size limit by doing
repeated history.pushState() calls. This would obviously take a while. To address the issue,
I am introducing a new window.internals API to override this limit. In the test, we now
lower this limit to 4MB, which makes the test run a lot faster.

* LayoutTests/loader/stateobjects/pushstate-size-iframe-expected.txt:
* LayoutTests/loader/stateobjects/pushstate-size-iframe.html:
* LayoutTests/loader/stateobjects/resources/pushstate-iframe.html:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/page/History.cpp:
(WebCore::History::totalStateObjectPayloadLimit const):
(WebCore::History::stateObjectAdded):
* Source/WebCore/page/History.h:
* Source/WebCore/page/LocalDOMWindow.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::resetToConsistentState):
(WebCore::Internals::setHistoryTotalStateObjectPayloadLimitOverride):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: https://commits.webkit.org/275452@main
